File: consensus/consensus.py
DNAstrings = {}

with open("rosalind_consensus.txt", "r") as f:
    label = ""
    while True:
        line = f.readline()
        if not line:
            break
        if(line[0] != '>'):
            DNAstrings[label] += line.strip()
        elif(line[0] == '>'):
            label = line[1:]
            DNAstrings[label] = ""


profile = [[0 for j in range(4)] for i in range(len(DNAstrings[label]))]

consensus = [0 for i in range(len(DNAstrings[label]))]

for label in DNAstrings:
    i = 0
    for base in DNAstrings[label]:
        if(base == "A"):
            profile[i][0] += 1
        elif(base == "C"):
            profile[i][1] += 1
        elif(base == "G"):
            profile[i][2] += 1
        elif(base == "T"):
            profile[i][3] += 1
        i += 1

options = ["A", "C", "G", "T"]

for i in range(len(profile)):
    greatest = 0
    for j in range(len(profile[i])):
        if(profile[i][j] > greatest):
            greatest = profile[i][j]
            consensus[i] = options[j]

print("".join(consensus))

A_profile = "A: "
C_profile = "C: "
G_profile = "G: "
T_profile = "T: "

for i in range(len(profile)):
    for j in range(len(profile[i])):
        if(j == 0):
            A_profile += str(profile[i][0]) + " "
        elif(j == 1):
            C_profile += str(profile[i][1]) + " "
        elif(j == 2):
            G_profile += str(profile[i][2])+ " "
        elif(j == 3):
            T_profile += str(profile[i][3])+ " "

print(A_profile)
print(C_profile)
print(G_profile)
print(T_profile)

File: consensus/consensus.cpp
#include <iostream> 
#include <fstream>
#include <string>
#include <vector>

using namespace std; 

typedef struct DNAStrand{
    string label;
    string sequence;
}DNAStrand;

int main(){
    ifstream f;
    int numStrands = 0;

    vector<DNAStrand> strands;

    f.open("rosalind_consensus.txt", ios::in);
    if(!f.is_open()){
        cout << "No file found";
        exit(1);
    }

    string line;

    while(getline(f, line)){
        if(line.at(0) == '>'){
            strands.insert(strands.end(), {line.substr(1), ""});
            numStrands++;
        } else {
            strands.at(numStrands-1).sequence.append(line);
        }
    }

    int size = strands.at(0).sequence.length();

    int **profile = (int **)malloc(4*sizeof(int *));
    for(int i = 0; i < 4; i++){
        profile[i] = (int *)calloc(size, sizeof(int));
    
    }

    char *consensus = (char *)calloc(size+1, sizeof(char));

    for(auto strand : strands){
        for(int i = 0; i < strand.sequence.size(); i++){
            if(strand.sequence.at(i) == 'A'){
                profile[i][0]++;
            } else if(strand.sequence.at(i) == 'C'){
                profile[i][1]++;
            } else if(strand.sequence.at(i) == 'G'){
                profile[i][2]++;
            } else if(strand.sequence.at(i) == 'T'){
                profile[i][3]++;
            }
        }
    }

    char options[] = {'A', 'C', 'G', 'T'};

    for(int m = 0; m < size; m++){
        int greatest = 0;
        for(int i = 0; i < 4; i++){
            if(profile[i][m] > greatest){
                greatest = profile[i][m];
                consensus[m] = options[i];
            }
        }
    }

    cout << consensus << endl;

    string Aprofile = "A: ";
    string Cprofile = "C: ";
    string Gprofile = "G: ";
    string Tprofile = "T: ";

    for(int m = 0; m < size; m++){
        int greatest = 0;
        for(int i = 0; i < 4; i++){
            if(i == 0){
                Aprofile += profile[m][i];
            } else if(i == 1){
                Cprofile += profile[m][i];
            } else if(i == 2){
                Gprofile += profile[m][i];
            } else if(i == 3){
                Tprofile += profile[m][i];
            }
        }
    }

    cout << Aprofile << endl;
    cout << Cprofile << endl;
    cout << Gprofile << endl;
    cout << Tprofile << endl;


    for(int i = 0; i < 4; i++){
        free(profile[i]);
    }

    free(profile);
    free(consensus);

    f.close();

    return 0;
}